File: csharp/src/Vista.SDK/UniversalIdBuilder.cs
using Vista.SDK.Internal;

namespace Vista.SDK;

public sealed partial record class UniversalIdBuilder : IUniversalIdBuilder
{
    public static readonly string NamingEntity = "data.dnv.com";
    private LocalIdBuilder? _localId;
    public ImoNumber? ImoNumber { get; private init; }

    public LocalIdBuilder? LocalId => _localId;

    public bool IsValid => ImoNumber is not null && _localId is not null && _localId.IsValid;

    public static UniversalIdBuilder Create(VisVersion version) =>
        new UniversalIdBuilder().WithLocalId(LocalIdBuilder.Create(version));

    public bool Equals(UniversalIdBuilder? other)
    {
        if (other is null)
            return false;
        return ImoNumber == other.ImoNumber && LocalId == other.LocalId;
    }

    public UniversalId Build()
    {
        return new UniversalId(this);
    }

    public UniversalIdBuilder WithLocalId(LocalIdBuilder localId) =>
        this with
        {
            _localId = localId
        };

    public UniversalIdBuilder WithoutLocalId() => this with { _localId = default };

    public bool TryWithLocalId(LocalIdBuilder? localId, out UniversalIdBuilder universalIdBuilder)
    {
        if (localId == null)
        {
            universalIdBuilder = this with { _localId = default };
            return false;
        }
        universalIdBuilder = this with { _localId = localId };
        return true;
    }

    public UniversalIdBuilder WithImoNumber(ImoNumber imoNumber) =>
        this with
        {
            ImoNumber = imoNumber
        };

    public UniversalIdBuilder WithoutImoNumber() => this with { ImoNumber = default };

    public UniversalIdBuilder TryWithImoNumber(ImoNumber? imoNumber)
    {
        if (imoNumber == null)
            return this;

        return this with
        {
            ImoNumber = imoNumber
        };
    }

    public bool TryWithImoNumber(ImoNumber? imoNumber, out UniversalIdBuilder universalIdBuilder)
    {
        if (imoNumber == null)
        {
            universalIdBuilder = this with { ImoNumber = default };
            return false;
        }
        universalIdBuilder = this with { ImoNumber = imoNumber };
        return true;
    }

    public sealed override int GetHashCode()
    {
        var hashCode = new HashCode();
        hashCode.Add(ImoNumber);
        hashCode.Add(LocalId);
        return hashCode.ToHashCode();
    }

    public override string ToString()
    {
        if (ImoNumber is null)
            throw new InvalidOperationException("Invalid Universal Id state: Missing IMO Number");
        if (LocalId is null)
            throw new InvalidOperationException("Invalid Universal Id state: Missing LocalId");

        string namingEntity = $"{NamingEntity}";
        using var lease = StringBuilderPool.Get();

        var builder = lease.Builder;

        builder.Append(namingEntity);
        builder.Append("/");
        builder.Append(ImoNumber.ToString());

        LocalId.ToString(builder);

        return lease.ToString();
    }
}

File: js/lib/UniversalId.Builder.ts
import { Codebooks, Gmod, LocalIdBuilder, UniversalId } from ".";
import { ImoNumber } from "./ImoNumber";
import { LocalIdParsingErrorBuilder } from "./internal/LocalIdParsingErrorBuilder";
import { Locations } from "./Location";
import { ParsingState } from "./types/LocalId";
import { UniversalIdParser } from "./UniversalId.Parsing";
import { VisVersion } from "./VisVersion";

export class UniversalIdBuilder {
    public static readonly namingEntity = "data.dnv.com";
    private _localId?: LocalIdBuilder;
    public imoNumber?: ImoNumber;

    public get localId() {
        return this._localId;
    }

    public get isValid(): boolean {
        return !!this.imoNumber?.isValid && !!this.localId?.isValid;
    }

    public static create(visVersion: VisVersion) {
        return new UniversalIdBuilder().withLocalId(
            LocalIdBuilder.create(visVersion)
        );
    }

    public equals(other?: UniversalIdBuilder): boolean {
        if (!other) return false;
        return (
            this.imoNumber === other.imoNumber &&
            !!this.localId?.equals(other.localId)
        );
    }

    public clone() {
        return Object.assign(new UniversalIdBuilder(), this);
    }

    public validate(
        errorBuilder = new LocalIdParsingErrorBuilder()
    ): LocalIdParsingErrorBuilder {
        if (!this.imoNumber?.isValid) {
            errorBuilder.push({
                type: ParsingState.EmptyState,
                message: "Missing or invalid IMO number",
            });
        }
        this._localId?.validate(errorBuilder);
        return errorBuilder;
    }

    public static parse(
        universalId: string,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.parse(
            universalId,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async parseAsync(
        universalIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.parseAsync(universalIdString, errorBuilder);
    }

    public static tryParse(
        universalId: string,
        gmod: Gmod,
        codebooks: Codebooks,
        locations: Locations,
        errorBuilder?: LocalIdParsingErrorBuilder
    ): UniversalIdBuilder | undefined {
        return UniversalIdParser.tryParse(
            universalId,
            gmod,
            codebooks,
            locations,
            errorBuilder
        );
    }

    public static async tryParseAsync(
        universalIdString: string | undefined,
        errorBuilder?: LocalIdParsingErrorBuilder
    ) {
        return UniversalIdParser.tryParseAsync(universalIdString, errorBuilder);
    }

    public build(): UniversalId {
        return new UniversalId(this);
    }

    public toString() {
        const builder: string[] = [];

        if (!this.imoNumber)
            throw new Error("Invalid Universal Id state: Missing IMO Number");
        if (!this.localId)
            throw new Error("Invalid Universal Id state: Missing LocalId");

        builder.push(UniversalIdBuilder.namingEntity);
        builder.push("/");
        builder.push(this.imoNumber.toString());
        this.localId.toString(builder);

        return builder.join("");
    }

    public tryWithLocalId(localId?: LocalIdBuilder) {
        if (!localId) return this;
        return this.with((s) => (s._localId = localId));
    }

    public withLocalId(localId: LocalIdBuilder) {
        return this.with((s) => (s._localId = localId));
    }

    public withoutLocalId() {
        return this.with((s) => (s._localId = undefined));
    }

    public withImoNumber(imoNumber: ImoNumber) {
        return this.with((s) => (s.imoNumber = imoNumber));
    }

    public tryWithImoNumber(imoNumber?: ImoNumber) {
        if (!imoNumber) return this;
        return this.with((s) => (s.imoNumber = imoNumber));
    }

    public withoutImoNumber() {
        return this.with((s) => (s.imoNumber = undefined));
    }

    public with(u: (state: UniversalIdBuilder) => void): UniversalIdBuilder {
        const n = this.clone();
        u && u(n);
        return n;
    }
}
